src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c

```diff
@@ -323,3 +387,659 @@ X509* CryptoNative_X509UpRef(X509* x509)
+    }
+
+    OCSP_CERTID_free(certId);
+    return ret;
+}
+
+static int Get0CertAndIssuer(X509_STORE_CTX* storeCtx, X509** subject, X509** issuer)
+{
+    assert(storeCtx != NULL);
+    assert(subject != NULL);
+    assert(issuer != NULL);
+
+    // get0 => don't free.
+    X509Stack* chain = X509_STORE_CTX_get0_chain(storeCtx);
+    int chainSize = chain == NULL ? 0 : sk_X509_num(chain);
+
+    if (chainSize < 1)
+    {
+        return 0;
+    }
+
+    *subject = sk_X509_value(chain, 0);
+    *issuer = sk_X509_value(chain, chainSize == 1 ? 0 : 1);
+    return 1;
+}
+
+static time_t GetIssuanceWindowStart()
+{
+    // time_t granularity is seconds, so subtract 4 days worth of seconds.
+    // The 4 day policy is based on the CA/Browser Forum Baseline Requirements
+    // (version 1.6.3) section 4.9.10 (On-Line Revocation Checking Requirements)
+    time_t t = time(NULL);
+    t -= 4 * 24 * 60 * 60;
+    return t;
+}
+
+X509VerifyStatusCode CryptoNative_X509ChainGetCachedOcspStatus(X509_STORE_CTX* storeCtx, char* cachePath)
+{
+    if (storeCtx == NULL || cachePath == NULL)
+    {
+        return (X509VerifyStatusCode)-1;
+    }
+
+    X509* subject;
+    X509* issuer;
+
+    if (!Get0CertAndIssuer(storeCtx, &subject, &issuer))
+    {
+        return (X509VerifyStatusCode)-2;
+    }
+
+    X509VerifyStatusCode ret = PAL_X509_V_ERR_UNABLE_TO_GET_CRL;
+    char* fullPath = BuildOcspCacheFilename(cachePath, subject);
+
+    if (fullPath == NULL)
+    {
+        return ret;
+    }
+
+    BIO* bio = BIO_new_file(fullPath, "rb");
+    OCSP_RESPONSE* resp = NULL;
+
+    if (bio != NULL)
+    {
+        resp = d2i_OCSP_RESPONSE_bio(bio, NULL);
+        BIO_free(bio);
+    }
+
+    if (resp != NULL)
+    {
+        ASN1_GENERALIZEDTIME* thisUpdate = NULL;
+        ASN1_GENERALIZEDTIME* nextUpdate = NULL;
+        ret = CheckOcsp(NULL, resp, subject, issuer, storeCtx, &thisUpdate, &nextUpdate);
+
+        if (ret != PAL_X509_V_ERR_UNABLE_TO_GET_CRL)
+        {
+            time_t oldest = GetIssuanceWindowStart();
+
+            // If either the thisUpdate or nextUpdate is missing we can't determine policy, so reject it.
+            // oldest = now - window;
+            //
+            // if thisUpdate < oldest || nextUpdate < now, reject.
+            //
+            // Since X509_cmp(_current)_time returns 0 on error, do a <= 0 check.
+            if (nextUpdate == NULL ||
+                thisUpdate == NULL ||
+                X509_cmp_current_time(nextUpdate) <= 0 ||
+                X509_cmp_time(thisUpdate, &oldest) <= 0)
+            {
+                ret = PAL_X509_V_ERR_UNABLE_TO_GET_CRL;
+            }
+        }
+
+        if (nextUpdate != NULL)
+        {
+            ASN1_GENERALIZEDTIME_free(nextUpdate);
+        }
+
+        if (thisUpdate != NULL)
+        {
+            ASN1_GENERALIZEDTIME_free(thisUpdate);
+        }
+    }
+
+    if (ret == PAL_X509_V_ERR_UNABLE_TO_GET_CRL)
+    {
+        unlink(fullPath);
```

Comment:
Why don't we want to always delete it?

Reply:
If it gave a useful answer (i.e. we changed the value of `ret`) the cache entry is still valuable.